From cb4f81c6f891e33cc278a8258a1170f1e71ae165 Mon Sep 17 00:00:00 2001
From: Robert Dabrowski <rdabrowski@maxlinear.com>
Date: Thu, 26 Nov 2020 20:13:26 +0100
Subject: [PATCH] PONRTSYS-8283: Enhance cpu packet trap handling

Trap if proto IP_P_ALL and MAC set(src or dest)
Fix default bSLAN_Vid
Trap all packets for ETH_P_ALL when bSubIfIdEnable is set
---
 drivers/net/datapath/pon_qos/pon_qos_tc_trap.c | 31 ++++++++++++++++++--------
 1 file changed, 22 insertions(+), 9 deletions(-)

diff --git a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
index e5960cff1848..548a84b28fb4 100644
--- a/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
+++ b/drivers/net/datapath/pon_qos/pon_qos_tc_trap.c
@@ -52,7 +52,7 @@ static void pon_qos_trap_set_default(struct net_device *dev,
 	pce_rule->pattern.bEnable = 1;
 	pce_rule->pattern.bInsertionFlag_Enable = 1;
 	pce_rule->pattern.nInsertionFlag = 0;
-	pce_rule->pattern.bSLAN_Vid = 1;
+	pce_rule->pattern.bSLAN_Vid = 0;
 	pce_rule->region = GSW_PCE_RULE_COMMMON;
 
 	pce_blk_info->region = pce_rule->region;
@@ -173,6 +173,12 @@ static void pon_qos_trap_parse_key_basic(struct net_device *dev,
 	return;
 }
 
+static bool is_mac_set(GSW_PCE_rule_t *pce_rule)
+{
+	return pce_rule->pattern.bMAC_SrcEnable ||
+	       pce_rule->pattern.bMAC_DstEnable;
+}
+
 static void pon_qos_trap_parse_key_vlan(struct net_device *dev,
 					const struct tc_cls_flower_offload *f,
 					GSW_PCE_rule_t *pce_rule)
@@ -182,19 +188,26 @@ static void pon_qos_trap_parse_key_vlan(struct net_device *dev,
 
 	if (dissector_uses_key(f->dissector, FLOW_DISSECTOR_KEY_VLAN)) {
 		key = skb_flow_dissector_target(f->dissector,
-						  FLOW_DISSECTOR_KEY_VLAN,
-						  f->key);
+						FLOW_DISSECTOR_KEY_VLAN,
+						f->key);
 		mask = skb_flow_dissector_target(f->dissector,
-						  FLOW_DISSECTOR_KEY_VLAN,
-						  f->mask);
-		if (!key && key->vlan_id && !mask && mask->vlan_id == 0xfff) {
+						 FLOW_DISSECTOR_KEY_VLAN,
+						 f->mask);
+
+		if (key && key->vlan_id && mask && mask->vlan_id == 0xfff) {
 			pce_rule->pattern.nSLAN_Vid = key->vlan_id;
+			pce_rule->pattern.bSLAN_Vid = 1;
 			return;
 		}
 	}
 
+	if (f->common.protocol == htons(ETH_P_ALL) &&
+	    (is_mac_set(pce_rule) || pce_rule->pattern.bSubIfIdEnable))
+		return;
+
 	pce_rule->pattern.nOuterVidRange = 4095;
 	pce_rule->pattern.bSVidRange_Select = 1;
+	pce_rule->pattern.bSLAN_Vid = 1;
 
 	if (eth_type_vlan(f->common.protocol)) {
 		pce_rule->pattern.bSLANVid_Exclude = 0;
@@ -377,12 +390,12 @@ int pon_qos_trap_offload(struct net_device *dev,
 	/* Set protocol based on ip_proto and ethertype based on n_proto */
 	pon_qos_trap_parse_key_basic(dev, f, pce_rule);
 
-	/* Set SLAN_Vid */
-	pon_qos_trap_parse_key_vlan(dev, f, pce_rule);
-
 	/* Set MAC source and destination */
 	pon_qos_trap_parse_key_eth(dev, f, pce_rule);
 
+	/* Set SLAN_Vid */
+	pon_qos_trap_parse_key_vlan(dev, f, pce_rule);
+
 	/* Set ICMP type */
 	pon_qos_trap_parse_key_icmp(dev, f, pce_rule);
 
